### PR TITLE
feat: upgrade to ES7

### PR DESCRIPTION
**Pull request by jywadhwani**
Tested by ingesting data using the ingest action method for datasets and corp users. Successfully ingested the data into GMS and was able to search for the same in DH UI. Also called the search endpoint to verify that the documents are indeed in search index

Request:
`curl http://localhost:9200/datasetdocument/_search -H 'Content-Type: application/json' -d '{"query": {"match_all": {}}}' | jq`

Response:
``{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "datasetdocument",
        "_type": "_doc",
        "_id": "urn%3Ali%3Adataset%3A%28urn%3Ali%3Adataplatform%3Akafka%2Csamplekafkadataset%2Cprod%29",
        "_score": 1,
        "_source": {
          "urn": "urn:li:dataset:(urn:li:dataPlatform:kafka,SampleKafkaDataset,PROD)",
          "browsePaths": [
            "/prod/kafka/samplekafkadataset"
          ],
          "origin": "PROD",
          "name": "SampleKafkaDataset",
          "platform": "kafka",
          "hasSchema": true
        }
      },
      {
        "_index": "datasetdocument",
        "_type": "_doc",
        "_id": "urn%3Ali%3Adataset%3A%28urn%3Ali%3Adataplatform%3Afoo%2Cbar%2Cprod%29",
        "_score": 1,
        "_source": {
          "urn": "urn:li:dataset:(urn:li:dataPlatform:foo,bar,PROD)",
          "hasOwners": true,
          "owners": [
            "jdoe",
            "datahub"
          ],
          "browsePaths": [
            "/prod/foo/bar"
          ],
          "origin": "PROD",
          "name": "bar",
          "platform": "foo",
          "upstreams": [
            "urn:li:dataset:(urn:li:dataPlatform:foo,barUp,PROD)"
          ],
          "hasSchema": true
        }
      }
    ]
  }
}``

Request:
`curl http://localhost:9200/corpuserinfodocument/_search -H 'Content-Type: application/json' -d '{"query": {"match_all": {}}}' | jq`

Response:
``
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "corpuserinfodocument",
        "_type": "_doc",
        "_id": "urn%3Ali%3Acorpuser%3Afbar",
        "_score": 1,
        "_source": {
          "urn": "urn:li:corpuser:fbar",
          "emails": [
            "fbar@linkedin.com"
          ],
          "ldap": "fbar",
          "managerLdap": "",
          "fullName": "Foo Bar",
          "active": true,
          "title": "",
          "skills": [],
          "teams": [],
          "aboutMe": ""
        }
      }
    ]
  }
}
``

Also verified by calling the following rest APIs
Search:
`curl "http://localhost:8080/datasets?q=search&input=bar" -X GET -H 'X-RestLi-Protocol-Version: 2.0.0' -H 'X-RestLi-Method: finder' | jq`
Typeahead:
`curl "http://localhost:8080/datasets?action=autocomplete" -d '{"query": "bar", "field": "name", "limit": 10, "filter": {"criteria": []}}' -X POST -H 'X-RestLi-Protocol-Version: 2.0.0' | jq`
Browse:
`curl "http://localhost:8080/datasets?action=browse" -d '{"path": "/prod", "start": 0, "limit": 10}' -X POST -H 'X-RestLi-Protocol-Version: 2.0.0' | jq`

## Checklist
- [X] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [X] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

**Comment by cobolbaby** [comment]
LGTM.

**Comment by keremsahin1** [comment]
This direct upgrade from 5 to 7 might be tricky. I would suggest you to test a scenario where there is an index created at 5.6 and then check if you can use that stored data with 7.

I faced a similar challenge for Neo4j 4.0 upgrade and Neo4j solves this problem very easily by just setting a flag by allowing inplace db upgrades. I don't know if ES provides something like that. It's worth to figure it. If not, we should probably help users by guiding them to migrate their ES data from 5.6 to 7.*. That's what we need to do internally at Linkedin anyways.

**Comment by jplaisted** [comment]
I agree with @keremsahin1 's comments, but I think it is fine to merge this into its own branch, as you've proposed already.

We should nail down the story before merging to master.

**Review by jplaisted — APPROVED** [review]

**Comment by camelliazhang** [comment]
Would you mind summarizing the major changes from 5 to 7 upgrade?

**Comment by jywadhwani** [comment]
> Would you mind summarizing the major changes from 5 to 7 upgrade?

Certainly, I will document the changes w.r.t. the index mapping/settings as well the write path.

**Comment by keremsahin1** [comment]
> I agree with @keremsahin1 's comments, but I think it is fine to merge this into its own branch, as you've proposed already.
> 
> We should nail down the story before merging to master.

Missed that it's a separate branch. That's fine then.

File: metadata-models/src/test/java/com/linkedin/metadata/ModelValidation.java [hunk 1]
```diff
@@ -48,16 +49,20 @@ public void validateSnapshots() throws Exception {
     getClassesInPackage("com.linkedin.metadata.snapshot", IGNORED_SNAPSHOT_CLASSES).forEach(
         SnapshotValidator::validateSnapshotSchema);
 
-    SnapshotValidator.validateUniqueUrn(
-        getClassesInPackage("com.linkedin.metadata.snapshot", IGNORED_SNAPSHOT_CLASSES).collect(Collectors.toList()));
+    @SuppressWarnings("unchecked")
```
Comment (keremsahin1) [inline metadata-models/src/test/java/com/linkedin/metadata/ModelValidation.java:52]: Is this change related to ES migration? Or, is it only a side effect of bumping to 0.2.13? Assuming it's the latter. Then, better bump master to 0.2.13 first and then branch off from there.

Comment (jywadhwani) [reply metadata-models/src/test/java/com/linkedin/metadata/ModelValidation.java:52]: Yes, it's the result of bumping up `datahub-gma` version to 0.2.13.
Considering the ES7 changes introduced in this PR will sit in this branch for some time before we merge to `master`, there will be changes in `master` branch corresponding to version bump of `datahub-gma`, so to some extent there will be merge conflicts before this PR is merged to `master`.
Was your concern around merge conflicts or restricting the PR to ES7 changes only?

**Comment by jywadhwani** [comment]
> Would you mind summarizing the major changes from 5 to 7 upgrade?

Added a markdown file summarizing the changes associated with the upgrade

File: metadata-models/src/test/java/com/linkedin/metadata/ModelValidation.java (hunk 1 above)
Comment (keremsahin1) [reply metadata-models/src/test/java/com/linkedin/metadata/ModelValidation.java:52]: The latter, restricting the PR to ES7 changes only.

File: li-utils/src/test/java/com/linkedin/common/urn/DatasetFieldUrnTest.java [hunk 2]
```diff
@@ -32,7 +32,7 @@ public void testSerialization() throws URISyntaxException {
   }
 
   @Test
-  public void testCreateUrn() {
+  public void testCreateUrn() throws URISyntaxException {
     final DatasetFieldUrn datasetFieldUrn = new DatasetFieldUrn(PLATFORM, DATASET_NAME, FABRIC_TYPE, FIELD_NAME);
```
Comment (camelliazhang) [inline li-utils/src/test/java/com/linkedin/common/urn/DatasetFieldUrnTest.java:36]: Does not seem to related to ES7 upgrade?

File: gms/impl/src/test/resources/corpUserESSearchRequest.json [hunk 3]
```diff
@@ -22,18 +22,17 @@
                     "fuzzy_transpositions" : true,
                     "lenient" : false,
                     "zero_terms_query" : "NONE",
+                    "auto_generate_synonyms_phrase_query": true,
```
Comment (camelliazhang) [inline gms/impl/src/test/resources/corpUserESSearchRequest.json:25]: I think this is optional?

File: gms/impl/src/test/resources/corpGroupESAutocompleteRequest.json [hunk 4]
```diff
@@ -7,7 +7,6 @@
   },
   "post_filter" : {
     "bool" : {
-      "disable_coord" : false,
```
Comment (camelliazhang) [inline gms/impl/src/test/resources/corpGroupESAutocompleteRequest.json:10]: These are auto-generated anyway?

File: li-utils/src/test/java/com/linkedin/common/urn/DatasetFieldUrnTest.java (hunk 2 above)
Comment (jplaisted) [reply li-utils/src/test/java/com/linkedin/common/urn/DatasetFieldUrnTest.java:36]: +1, also these methods don't actually throw URISyntaxException? This isn't needed?

File: metadata-models/src/test/java/com/linkedin/metadata/ModelValidation.java (hunk 1 above)
Comment (jywadhwani) [reply metadata-models/src/test/java/com/linkedin/metadata/ModelValidation.java:52]: Sure, we can upgrade `datahub-gma` version and restrict this PR to ES7 changes only.

File: docs/advanced/es-7-upgrade.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Elasticsearch upgrade from 5.6.8 to 7.9.3: Summary of changes
+
+### Search index mapping & settings
+- Removal of mapping types (as mentioned [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html))
+- Specify the maximum allowed difference between `min_gram` and `max_gram` for NGramTokenizer and NGramTokenFilter by adding property `max_ngram_diff` in index settings, particularly if the difference is greater than 1 (as mentioned [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html))
+
+
+### Search query
+- `disable_coord` parameter of the `bool` and `common_terms` queries has been removed (as mentioned [here](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/breaking-changes-6.0.html))
+- `auto_generate_synonyms_phrase_query` parameter in `match` query is added with a default value of `true` (as mentioned [here](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/query-dsl-match-query.html))
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply docs/advanced/es-7-upgrade.md:10]: Yes they are auto-generated but I think it is worth it calling out. Particularly if users have specified tests where they expect search query with certain parameters - those parameters will change when we move to ES 7.x.
I will mention that these are optional parameters and automatically populated in search query.

File: docs/advanced/es-7-upgrade.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+### Java High Level Rest Client
+- In 7.9.3, Java High Level Rest Client instance needs a REST low-level client builder to be built. In 5.6.8, the same instance needs REST low-level client
+- Document APIs such as the Index API, Delete API, etc no longer takes the doc `type` as an input
```
[the remark this answers is not in the document]
Comment (jywadhwani) [reply docs/advanced/es-7-upgrade.md:14]: `ESSearchDAO` PR is in `datahub-gma` https://github.com/linkedin/datahub-gma/pull/43
Particularly commit https://github.com/linkedin/datahub-gma/pull/43/commits/b27596d8f3c62b2198a795f071e0d0ad00e2f539

File: gms/impl/src/test/resources/corpGroupESAutocompleteRequest.json (hunk 4 above)
Comment (jywadhwani) [reply gms/impl/src/test/resources/corpGroupESAutocompleteRequest.json:10]: Yes. It's worth calling out in the migration summary doc, I think.

File: gms/impl/src/test/resources/corpUserESSearchRequest.json (hunk 3 above)
Comment (jywadhwani) [reply gms/impl/src/test/resources/corpUserESSearchRequest.json:25]: Yes, and auto-generated

File: li-utils/src/test/java/com/linkedin/common/urn/DatasetFieldUrnTest.java (hunk 2 above)
Comment (jywadhwani) [reply li-utils/src/test/java/com/linkedin/common/urn/DatasetFieldUrnTest.java:36]: As mentioned in the other thread with Kerem, I agree that it is better to restrict ES7 upgrade changes to this PR.

**Comment by cobolbaby** [comment]
Any progress on this issue? 

Once the request is merged, I plan to migrate to the cluster maintained by the es administrator.

**Comment by jplaisted** [comment]
@cobolbaby a few things:

1) Notice this PR is _not_ to master. We're going to merge into an ES7 branch for awhile, until we can add integration tests and do a little bit of battle testing. We also want to write a migration plan for folks that are currently using ES5 like yourself. Then we'll merge to master (essentially _removing_ 5 support). I'd imagine this may be done by / in January. If you want to use this branch early, you're more than welcome to, but I don't think we're going to publish docker images and stuff for it.
2) @jywadhwani will be on leave until March. @camelliazhang or myself will need to take this PR over.

**Comment by cobolbaby** [comment]
There is no problem in merging to a non-master branch. The GMS container image we currently use is also self-compiled.

As long as the core functions are not affected, I am willing to have a try, even if the unit test is not perfect.

File: docs/advanced/es-7-upgrade.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# Elasticsearch upgrade from 5.6.8 to 7.9.3: Summary of changes
+
+### Search index mapping & settings
+- Removal of mapping types (as mentioned [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html))
+- Specify the maximum allowed difference between `min_gram` and `max_gram` for NGramTokenizer and NGramTokenFilter by adding property `max_ngram_diff` in index settings, particularly if the difference is greater than 1 (as mentioned [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html))
```
[the remark this answers is not in the document]
Comment (jplaisted) [reply docs/advanced/es-7-upgrade.md:5]: I see this when copying over settings / mappings, but I'm not sure what is triggering it specifically.

Using python:

```
elasticsearch.exceptions.RequestError: RequestError(400, 'illegal_argument_exception', 'The difference between max_gram and min_gram in NGram Tokenizer must be less than or equal to: [1] but was [47]. This limit can be set by changing the [index.max_ngram_diff] index level setting.')
```

I guess we have this tokenizer that is causing it

```
                        "custom_ngram": {
                            "type": "ngram",
                            "min_gram": "3",
                            "max_gram": "50"
                        },
```

[the remark this answers is not in the document]
Comment (jplaisted) [reply docs/advanced/es-7-upgrade.md:5]: So needless to say the settings and mappings need to be updated in this PR.

**Comment by jplaisted** [comment]
See https://github.com/linkedin/datahub/pull/2092
